### PR TITLE
Refactors bar drink icons into datum singletons / unit tests them

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

- Refactors bar drink icons.
   - Juice boxes no longer have a hard-coded list of a bunch of reagent types in their update state, and use a system similar to bar drinks.
   - Glass and shot glass icon information are no longer stored on the drink. Instead, they are now stored in glass style datums. These datums store name, description, icon, and icon state of a certain container + reagent type.
   - Glass styles are applied via the `takes_reagent_appearance` component. Glasses, shot glasses, and juice boxes have this component. 
   - This comes with support for being able to have drink icons from different files, rather than requiring the drinks DMI.
- The britmug is now a subtype of mug. 
   -  1 new icon: britmug filled.
- Various small code clean-up around drink reagents. 
- Unit tests icon state setups for glass styles as well as all `/drink` reagent container subtypes. 
- Splits up the massive `drinks.dmi` into separate files. 

*Disclaimer: Much of the drinking glass datums were written via script automatically, so there may be errors present.*

## Why It's Good For The Game

- Much easier to add new drink styles, much more modular. 
   - It is no longer necessary for new drinks to be added to the massive `drinks.dmi`. People working with drinks in the future can simply add their glass style datum and point it to their file wherever it may be. 
- Expandable system. Adding a new type of reagent container that works similarly to bar drinks but for different types of icons is a breeze. 
- Ensures going forward no bar drinks have invisible sprites. 

## Changelog

:cl: Melbert
refactor: Refactored how bar drinks set their icons. Juice boxes now use the same system.
/:cl:

### CODE REVIEW

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 1]
```diff
@@ -1603,21 +1603,28 @@
...
 	color = "#50531a" // rgb: 175, 175, 0
 	taste_description = "horribly stringy, bitter pulp"
-	glass_name = "glass of banana peel pulp"
-	glass_desc = "Ancient Clown Lore says that pulped banana peels are good for your blood, but are you really going to take medical advice from a clown about bananas?"
 	metabolization_rate = 0.5 * REAGENTS_METABOLISM
 	clot_rate = 0.2
 	passive_bleed_modifier = 0.8
 	chemical_flags = REAGENT_CAN_BE_SYNTHESIZED
 
+/datum/glass_style/drinking_glass/seraka_extract
```
Comment (ZephyrTFA) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm:1611]: Should this be `banana_peel`?

File: code/datums/components/takes_reagent_appearance.dm [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		item_parent.icon_state = icon_state_pre_change
+	else if(style.icon && style.icon_state)
+		// style
+		item_parent.icon = style.icon
+		item_parent.icon_state = style.icon_state
+		on_icon_changed?.Invoke(style)
+		return COMSIG_ATOM_NO_UPDATE_ICON_STATE
+
+	// Reset gets invoked regardless, as further updates may "reset" the icon yet
+	on_icon_reset?.Invoke()
```
Comment (Mothblocks) [inline code/datums/components/takes_reagent_appearance.dm:137]:
```suggestion
	on_icon_reset?.InvokeAsync()
```

We're in a signal, so we cannot yield

File: code/modules/food_and_drinks/restaurant/generic_venues.dm [hunk 3]
```diff
@@ -83,20 +83,28 @@
...
-		glass_visual = initial(reagent_to_order.glass_icon_state)
-	else if(initial(reagent_to_order.shot_glass_icon_state))
-		glass_visual = initial(reagent_to_order.shot_glass_icon_state)
-	else if(initial(reagent_to_order.fallback_icon_state))
-		glass_visual = initial(reagent_to_order.fallback_icon_state)
-	else
-		stack_trace("[reagent_to_order] has no icon sprite for restaurant code, please set a fallback_icon_state for this reagent.")
+	// If we have no icon or fallback, well... Scream. Someone should fix it
+	if(!glass_visual_icon || !glass_visual_icon_state)
+		stack_trace("[reagent_to_order] has no icon sprite for restaurant code, please set a fallback icon for this reagent.")
```
Comment (Mothblocks) [inline code/modules/food_and_drinks/restaurant/generic_venues.dm:104]: Do we want to CRASH here instead? Or at least return a blank image?

File: code/modules/unit_tests/drink_icons.dm [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	for(var/container_type in GLOB.glass_style_singletons)
+		for(var/reagent_type in GLOB.glass_style_singletons[container_type])
+			var/datum/glass_style/style = GLOB.glass_style_singletons[container_type][reagent_type]
+			var/style_icon = style.icon
+			var/style_icon_state = style.icon_state
+
+			if(!style_icon_state)
+				continue
+			if(!style_icon)
+				Fail("[style.type] glass style had an icon state ([style_icon_state]) but no icon file.")
```
Comment (Mothblocks) [inline code/modules/unit_tests/drink_icons.dm:18]:
```suggestion
				TEST_FAIL("[style.type] glass style had an icon state ([style_icon_state]) but no icon file.")
```

File: code/modules/unit_tests/drink_icons.dm [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+			// to see if it's just misplaced and the user needs to just correct it
+			if(style_icon != generic_mixed_drink_loc && icon_exists(generic_mixed_drink_loc, style_icon_state))
+				was_actually_in = "The icon was found in the mixed drinks dmi."
+			else if(style_icon != generic_drink_loc && icon_exists(generic_drink_loc, style_icon_state))
+				was_actually_in = "The icon was found in the standard drinks dmi."
+			// If it wasn't found in either of the generic spots it could be absent or otherwise in another file
+			else
+				was_actually_in = "The icon may be located in another dmi or is missing."
+
+			Fail("[style.type] glass style had an icon state ([style_icon_state]) not present in its icon ([style_icon]). [was_actually_in]")
```
Comment (Mothblocks) [inline code/modules/unit_tests/drink_icons.dm:34]:
```suggestion
			TEST_FAIL("[style.type] glass style had an icon state ([style_icon_state]) not present in its icon ([style_icon]). [was_actually_in]")
```

File: code/modules/unit_tests/drink_icons.dm [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+/datum/unit_test/drink_icons
+
+/datum/unit_test/drink_icons/Run()
+	for(var/obj/item/reagent_containers/cup/glass/glass_subtypes as anything in subtypesof(/obj/item/reagent_containers/cup))
+		var/glass_icon = initial(glass_subtypes.icon)
+		var/glass_icon_state = initial(glass_subtypes.icon_state)
+		if(!glass_icon_state)
+			continue
+		if(!glass_icon)
+			Fail("[glass_subtypes] had an icon state ([glass_icon_state]) but no icon file.")
```
Comment (Mothblocks) [inline code/modules/unit_tests/drink_icons.dm:71]:
```suggestion
			TEST_FAIL("[glass_subtypes] had an icon state ([glass_icon_state]) but no icon file.")
```

File: code/modules/unit_tests/drink_icons.dm [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+			continue
+		if(icon_exists(glass_icon, glass_icon_state))
+			continue
+		Fail("[glass_subtypes] had an icon state ([glass_icon_state]) not present in its icon ([glass_icon]).")
```
Comment (Mothblocks) [inline code/modules/unit_tests/drink_icons.dm:75]:
```suggestion
		TEST_FAIL("[glass_subtypes] had an icon state ([glass_icon_state]) not present in its icon ([glass_icon]).")
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by MTandi** [comment]
@MrMelbert there is a bunch of recipes in the crafting menu that just turn glass bottle into the stylized bottles. Can we reuse this system to remove those recipes? Like when the glass changes the style, bottle will also adapt to the first drink you put into it.

**Comment by MTandi** [comment]
> what, there is? which recipes? I had no idea, but theoretically yes bottles can use this too

The drinks category of the crafting menu:

image

**Comment by MTandi** [comment]
@MrMelbert FYI, it seems like you implemented here a similar thing to my `default_container` for reagents in https://github.com/tgstation/tgstation/pull/71779

Although, you added it to have an icon for reagent, while I added a reference to a default container, it has similar functionality - I take the icon of the default container to visualize reagent in the crafting menu.

**Pull request merged**